### PR TITLE
feat(checkout): CHECKOUT-2368 Display Custom Shipping Method

### PR DESCRIPTION
**Pull request by bc-peng**
## What?
Display custom shipping method created by manual order UI on the storefront checkout page.

## Why?
Support the use of a custom shipping method.

### When the experiment is ON
Opening a draft order with a custom shipping method will jump to the payment step directly.

https://github.com/bigcommerce/checkout-js/assets/88361607/78e2ac4b-1245-41d4-9897-d39b820e6cc3

### When the experiment is OFF
The edit button of the shipping step appears on the page.

image

@bigcommerce/team-checkout

### CODE REVIEW

File: packages/core/src/app/checkout/Checkout.test.tsx [hunk 1]
```diff
@@ -255,6 +258,20 @@ describe('Checkout', () => {
...
 
+        it('renders custom shipping method and locks shipping component', async () => {
+            checkout.use('CartWithCustomShippingAndBilling');
+
+            render(<CheckoutTest {...defaultProps} />);
+
+            await checkout.waitForPaymentStep();
+
+            expect(screen.getByText('Manual Order Custom Shipping Method')).toBeInTheDocument();
+            expect(screen.getAllByRole('button', { name: 'Edit' })).toHaveLength(2);
```
Comment (bc-peng) [inline packages/core/src/app/checkout/Checkout.test.tsx:269]: We are asserting these two `Edit` buttons here.

image

File: packages/core/src/app/checkout/getCheckoutStepStatuses.ts [hunk 2]
```diff
@@ -180,12 +180,18 @@ const getShippingStepStatus = createSelector(
             cart && consignments ? hasUnassignedLineItems(consignments, cart.lineItems) : true;
         const isComplete = hasAddress && hasOptions && !hasUnassignedItems;
         const isRequired = itemsRequireShipping(cart, config);
+        const isCompleteWithCustomShipping =
```
Comment (animesh1987) [inline packages/core/src/app/checkout/getCheckoutStepStatuses.ts:183]: Can we please just call it `hasCustomShippingSelected`

File: packages/core/src/app/common/utility/isCustomShippingEnabled.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { CheckoutSettings } from '@bigcommerce/checkout-sdk';
+
+export default function isCustomShippingEnabled(
```
Comment (animesh1987) [inline packages/core/src/app/common/utility/isCustomShippingEnabled.ts:3]: Can we refactor this and have it generic so we can pass an experiment name and fallback is `true`.

File: packages/core/src/app/shipping/hasSelectedShippingOptions.ts [hunk 4]
```diff
@@ -9,12 +9,14 @@ export default function hasSelectedShippingOptions(consignments: Consignment[]):
...
-            ).length,
+            (consignment.selectedShippingOption &&
+                consignment.selectedShippingOption.id &&
+                // Selected option is available
+                consignment.availableShippingOptions &&
+                consignment.availableShippingOptions.filter(
+                    ({ id }) => id === consignment.selectedShippingOption?.id,
+                ).length) ||
+            (consignment.selectedShippingOption?.type === 'custom' &&
+                consignment.selectedShippingOption.cost >= 0),
```
Comment (animesh1987) [inline packages/core/src/app/shipping/hasSelectedShippingOptions.ts:20]: Don't think we need the cost comparison.

Comment (bc-peng) [reply packages/core/src/app/shipping/hasSelectedShippingOptions.ts]: Thanks for the feedback.
All comments have been addressed in the last commit, https://github.com/bigcommerce/checkout-jas/pull/1900/commits/b83b798c9769fe3c9ec5e5d87cc5270312bda14a.

**Review by animesh1987 — APPROVED** [review]

**Pull request merged**